presto-hive/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java

```diff
@@ -630,7 +631,7 @@ public synchronized void alterPartition(String databaseName, String tableName, P
                     childPartitionValues = listPartitions(fileStatus.getPath(), partitionColumns.subList(1, partitionColumns.size()));
                 }
 
-                String value = fileStatus.getPath().getName().substring(directoryPrefix.length());
+                String value = unescapePathName(fileStatus.getPath().getName().substring(directoryPrefix.length()));
```

Comment:
It is unclear to me what is the current semantics of `ExtendedHiveMetastore.listPartitions` concerning unescaping partition names.
I would guess that it is expected to return escaped values, which are then unescaped in `HivePartitionManager.
This semantics looks weird and I would like `ExtendedHiveMetastore` to not use escaped strings anywhere as an interface. Yet it seems to me that change here should imply removing escaping in `HivePartitionManager`.

Please talk to @haozhun how to straighten this up.

Reply:
Escaping in this context is not a problem that I have carefully thought about.

It is not possible that `ExtendedHiveMetastore` never use escaped strings. If you have a partition name, it has to be escaped because you could have `=` and `/` in partition column name or value.

Therefore, it seems to me that it would be reasonable if
* partitionName (`col1=x/col2=y`) is always escaped
* partitionColumn (`col1`, `col2`) and partitionValue (`x`, `y`) is never escaped.

Do you think this is reasonable and maintenable?

Reply:
I'm not sure I understand the suggestion.  The problem I was seeing is that with the FileHiveMetastore the Partition names were being escaped twice.  It looks like the other escaping comes from the makePartName() call on line 603 which does its own escaping.  So the two options I see are 
1. unescape before calling that method, which is what I do here
2. reimplement makePartName in FileHiveMetastore to do the same thing but without escaping the partition names.  I'm less inclined to do that since the MetastoreUtil method calls `makePartName` from hive's FileUtils (which is the bit that does the escaping), and even though the rest of the logic of that method probably won't change, in theory it could.

Reply:
but maybe I should do the unescaping higher up (i.e. in getPartitionNames() instead of listPartitions()) and put a comment explaining the reason

Reply:
> Therefore, it seems to me that it would be reasonable if
> 
> * partitionNames (col1=x/col2=y) are always escaped
> * partitionColumns (col1, col2) and partitionValues (x, y) are never escaped.
>
> Do you think this is reasonable and maintainable?
 
@Haozhun it would be great if we could remove `partitionNames` at all from being part of `EHM` interface. And always return and take as parameters parsed structures. Like list of `partitionValues` or list of tuples `(partitionKey, partitionValue)`. And perform encoding/decoding to/from `partitionNames` only internally.
 I do not know if that is possible and if how much of an effort is that. 

If above is not possible, then is seems to me that what you propose is reasonable. Isn't that what we have already?